### ISSUE TITLE
Type parameters declared in disabled context should have oblivious nullability

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#29980]
```
        [Fact]
        public void Constraints_65()
        {
            var source =
@"
#nullable disable 
interface I1
{
    void F1<TF1>();
}

#nullabe enable 
class A : I1
{
    public void F1<TF1A>() where TF1A : object
    {}
}
";
            var comp = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
            // PROTOTYPE(NullableReferenceTypes): unexpected warning
            comp.VerifyDiagnostics(
                // (21,17): warning CS8633: Nullability in constraints for type parameter 'TF1A' of method 'A.F1<TF1A>()' doesn't match the constraints for type parameter 'TF1' of interface method 'I1.F1<TF1>()'. Consider using an explicit interface implementation instead.
                //     public void F1<TF1A>() where TF1A : object
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInConstraintsOnImplicitImplementation, "F1").WithArguments("TF1A", "A.F1<TF1A>()", "TF1", "I1.F1<TF1>()").WithLocation(21, 17)
                );
        }
```

**Comment by AlekseyTs** [comment]
Another scenario:
```
        [Fact]
        public void Constraints_66()
        {
            var source =
@"
interface I1
{
    void F2<TF2>() where TF2 : object;
}

[System.Runtime.CompilerServices.NonNullTypes(false)]
class A : I1
{
    public void F2<TF2A>()
    {}
}
";
            var comp = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
            // https://github.com/dotnet/roslyn/issues/29980: unexpected warning
            comp.VerifyDiagnostics(
                // (23,17): warning CS8633: Nullability in constraints for type parameter 'TF2A' of method 'A.F2<TF2A>()' doesn't match the constraints for type parameter 'TF2' of interface method 'I1.F2<TF2>()'. Consider using an explicit interface implementation instead.
                //     public void F2<TF2A>()
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInConstraintsOnImplicitImplementation, "F2").WithArguments("TF2A", "A.F2<TF2A>()", "TF2", "I1.F2<TF2>()").WithLocation(23, 17)
                );
        }
```

**Comment by jcouv** [comment]
@AlekseyTs Is this issue still applicable?

**Comment by jaredpar** [comment]
This is still relevant and `Constraints_65` has an example that demonstrates the problem. In the case where it's explicitly constrained to `object` it's oblivious but when it's implicit it's treated differently.

**Comment by jaredpar** [comment]
Dug into this issue a bit and here is a repro that demonstrates where the compiler bug is: 

``` csharp
#nullable disable
interface I1 { 
  void M1<T1>(); 
  void M2<T2>() where T2 : object; 
}

The compiler treats them differently today and it can be demonstrated with the following:

#nullable enable
class C : I1 {
  public void M1<U1>() { } 
  public void M2<U2>() where U2 : object { } 
}
```
The compiler will end up issuing a warning here for `C.M1` but not for `C.M2`. The reason is that when there is no constraint on a type parameter the compiler fails to properly record the context in which the type parameter was defined. Instead of being oblivious this is instead marked as non-nulllable and hence a warning is issued for the `M1` override. For `M2` though the compiler can deduce it's oblivious due to the `object` constraint. 

After some discussion one possible approach to this is to insert an implicit `object~` constraint when no such one exists. That would allow us to properly record the oblivious nature and make it consistent across all our scenarios. That would likely require some metadata import work as well. 

One other issue we'd need to settle is what token settles what context the type parameter is declared in. If there is a `new()` constraint for instance do we use that to determine nullability context?

**Comment by jaredpar** [comment]
Make sure to include the test scenarios from #30214 when resolving this bug.

**Comment by gafter** [comment]
In my opinion, the warning produced for https://github.com/dotnet/roslyn/issues/29980#issue-361084142 is correct and should remain.  The correct fix is to remove the constraint on the implementation or, if you really want to change the constraint, suppress the warning.

In my opinion the analysis at https://github.com/dotnet/roslyn/issues/29980#issuecomment-461528268 is incorrect.  A constraint of `object` is not the same as the absence of a constraint, even in nullable disabled code, and should not be treated as the absence of a constraint.  By [previous LDM decisions](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-08-09.md#unconstrained-generics), the absence of a constraint is equivalent to an `object?` constraint.

In my opinion, the warning produced for https://github.com/dotnet/roslyn/issues/29980#issuecomment-422207167 is incorrect simply because we should never be producing nullability diagnostics for locations where nullability is disabled.  This is a compiler bug and should be fixed by properly suppressing nullability diagnostics in disabled regions.

**Comment by gafter** [comment]
After discussing this with @AlekseyTs and @MadsTorgersen offline, I think I understand the intent better, and have prepared https://github.com/dotnet/csharplang/issues/2398 to trigger LDM discussion of the proposed change.

There is a separate issue that we don't always suppress nullable warnings in `#nullable disable`d code, but that is orthogonal to this issue.
